metricbeat/beater/reload.go

```diff
@@ -0,0 +1,252 @@
+package beater
+
+import (
+	"expvar"
+	"fmt"
+	"os"
+	"path/filepath"
+	"sync"
+	"time"
+
+	"github.com/pkg/errors"
+
+	"github.com/elastic/beats/libbeat/common"
+	"github.com/elastic/beats/libbeat/logp"
+	"github.com/elastic/beats/libbeat/publisher"
+	"github.com/elastic/beats/metricbeat/mb"
+)
+
+var (
+	debugr        = logp.MakeDebug("reloader")
+	configReloads = expvar.NewInt("metricbeat.config.reloads")
+	moduleStarts  = expvar.NewInt("metricbeat.config.module.starts")
+	moduleStops   = expvar.NewInt("metricbeat.config.module.stops")
+	moduleRunning = expvar.NewInt("metricbeat.config.module.running")
+
+	ErrNoNewFiles = errors.New("no new files found")
+)
+
+type ConfigReloader struct {
+	registry RunningRegistry
+	config   ReloaderConfig
+	client   func() publisher.Client
+	done     chan struct{}
+	wg       sync.WaitGroup
+}
+
+type RunningRegistry struct {
+	List  map[uint64]ModuleRunner
+	mutex sync.Mutex
+}
+
+func NewConfigReloader(config ReloaderConfig, p publisher.Publisher) ConfigReloader {
+
+	// TODO: Is there any downside to call connect for each new module?
+	client := func() publisher.Client { return p.Connect() }
+
+	return ConfigReloader{
+		registry: NewRunningRegistry(),
+		config:   config,
+		client:   client,
+		done:     make(chan struct{}),
+	}
+}
+
+func (r *ConfigReloader) Start() {
```

Comment:
`s/Start/Run`? This method implements the run-loop itself. Using `Start` implies a go-routine being started, but returns immediately, `Run` implies the method blocks.